### ISSUE TITLE
clarity on use of tskit.union()

### ISSUE THREAD

**Issue by ShyamieG** [tskit-dev/tskit#2905]
Hello!

I'm not experiencing an issue per se, but rather hoping to get some clarification of what is going on under the hood with tskit.union(). In particular, I want to understand exactly what 'check_shared_equality' is doing/looking for. I poked around the source code a bit, but wasn't able to find a definitive answer.

I have [two trees](https://github.com/tskit-dev/tskit/files/14284629/tsfiles.zip) generated by SLiM that share 15 nodes in common. When I subset the trees on these 15 nodes, their topology is identical. I verified this by eye, labelling the nodes by their slim_ids:

Tree 1
```
33.00┊  6   8 10 12 14 28 30 36 24 38 ┊    
     ┊  ┃   ┃  ┃                      ┊    
32.00┊ 94  80 84                      ┊    
     ┊  ┃                             ┊    
31.00┊ 244                            ┊    
     ┊  ┃                             ┊    
30.00┊ 510                            ┊    
     0                            24214675 
```

Tree2
```
33.00┊  6   8 10 12 14 28 30 36 24 38 ┊    
     ┊  ┃   ┃  ┃                      ┊    
32.00┊ 94  80 84                      ┊    
     ┊  ┃                             ┊    
31.00┊ 244                            ┊    
     ┊  ┃                             ┊    
30.00┊ 510                            ┊    
     0                            24214675 
```

When I try to merge these trees with tskit.union() and the flag 'check_shared_equality = True', I get the following error:

```
_tskit.LibraryError: Shared portions of the tree sequences are not equal. (TSK_ERR_UNION_DIFF_HISTORIES)
```

The trees can be successfully merged when I set 'check_shared_equality = False'. I checked the resulting merged tree manually, and everything looks as I would expect it to (see attached text files).

So what is 'check_shared_equality' looking for? I noticed that the 'node' and 'individual' tables are not identical when I subset tree1 and tree2 on their shared nodes - they are not in the same order, and for the 'individual' tables, the flags are not consistent. This is not a problem for my specific case, but I just wanted to verify that this is in fact what 'check_shared_equality' is actually 'checking'. Relatedly, if two trees can be successfully merged when 'check_shared_equality = False', **is it safe to assume that the topology of their shared nodes is identical**?

To summarize, my current understanding is that with 'check_shared_equality = True', tskit.union() is looking for ALL the table collections to be identical across subsets, while if 'check_shared_equality = False' only the edge tables need to be identical. Is this about right?

I have attached drawings of tree1, tree2, and the merged tree [here](https://github.com/tskit-dev/tskit/files/14283963/trees.zip) in case that is useful. Thanks as always for any help!

**Comment by petrelharp** [comment]
Yes, good question. What that check does is to [subset](https://tskit.dev/tskit/docs/latest/python-api.html#tskit.TreeSequence.subset) each of the tree sequences on the nodes that should be the same, and then check that the resulting subsets are identical. Subsetting
> will retain only the nodes whose IDs are listed in nodes, only edges for which both parent and child are in nodes`, only mutations whose node is in nodes, and only individuals that are referred to by one of the retained nodes. 
(However, as you can see from the [place in the code where this happens](https://github.com/tskit-dev/tskit/blob/1601a5e09b68191d062e52c1a0dcec093149e1c1/c/tskit/tables.c#L13157) there's a bit more going on: to check for identical-ness we need to do some potential reordering, which is why the docs are more clear on what exactly happens.)

My first guess is that what's going on is that something like the node flags are different? For instance, if something's a sample in one tree but not the other, that can happen.

Hm, but looking into your example, I'm not sure. Here's a quick pass looking at this (maybe with bugs, beware!!):
```
ts1 = tskit.load("tree1.ts")
ts2 = tskit.load("tree2.ts")
sid1 = ts1.tables.nodes.metadata_vector("slim_id")
sid2 = ts2.tables.nodes.metadata_vector("slim_id")
both = np.intersect1d(sid1, sid2)
n1 = np.array([np.where(sid1 == a)[0][0] for a in both])
n2 = np.array([np.where(sid2 == a)[0][0] for a in both])
sub1 = ts1.subset(n1)
sub2 = ts2.subset(n1)
subt1 = sub1.dump_tables()
subt2 = sub2.dump_tables()
subt1.canonicalise()
subt2.canonicalise()

sub1.tables.assert_equals(sub2.tables, ignore_metadata=True)
```
which gets
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/peter/projects/tskit-dev/treestuff_venv/lib/python3.11/site-packages/tskit/tables.py", line 3255, in assert_equals
    table.assert_equals(
  File "/home/peter/projects/tskit-dev/treestuff_venv/lib/python3.11/site-packages/tskit/tables.py", line 455, in assert_equals
    raise AssertionError(
AssertionError: EdgeTable row 0 differs:
self.parent=13 other.parent=1
self.child=14 other.child=7
```
This suggests that I have a bug above. But if not, maybe it suggests that the trees aren't actually the same? Anyhow, it should get you on the right track.  (Note that I've used `tables.assert_equals` since it tells us what's different in a nice way.)

**Comment by ShyamieG** [comment]
Thank you, this is super helpful! I hadn't come across assert_equals before - using that would have saved me a lot of time 😂 

The bug in your code was here:

```
sub2 = ts2.subset(n1) # should be n2
```

When I correct that, your code gives:

```
AssertionError: IndividualTable row 0 differs:
self.flags=196608 other.flags=131072
self.metadata={'pedigree_id': 3, 'pedigree_p1': -1, 'pedigree_p2': -1, 'age': 16, 'subpopulation': 1, 'sex': -1, 'flags': 0} other.metadata={'pedigree_id': 3, 'pedigree_p1': -1, 'pedigree_p2': -1, 'age': 17, 'subpopulation': 1, 'sex': -1, 'flags': 0}
```

I don't care about flag mismatches, and this doesn't seem to prevent tree merging (as long as check_shared_equality is False).

It seems like this code will only show one discrepancy at a time - if that is true, is there a way to see all conflicts between two trees at once?

**Comment by ShyamieG** [comment]
Playing more with this, the node tables are not equivalent because the individual id - slim id pairings are not the same across both trees. Makes sense!

Thanks again :D

**Comment by petrelharp** [comment]
Ah, that makes sense - that's the sort of thing I would have expected. And, no - there isn't a single method that gives all differences; that would be useful in this case.
